apps/indexer/src/services/events/event-outbox-fan-out-service.ts

```diff
@@ -0,0 +1,142 @@
+import { sql } from "drizzle-orm";
+import Deferred from "promise-deferred";
+import { schema } from "../../../schema.ts";
+import type { DB } from "../db.ts";
+
+type EventOutboxFanOutService_FanOutEventsParams = {
+  signal: AbortSignal;
+};
+
+type EventOutboxFanOutServiceOptions = {
+  db: DB;
+  /**
+   * Polling interval in milliseconds for checking for new events.
+   * This works as fallback to LISTEN/NOTIFY if there are no new events.
+   *
+   * @default 30_000
+   */
+  pollInterval?: number;
+};
+
+export class EventOutboxFanOutService {
+  private readonly db: DB;
+  private readonly pollInterval: number;
+  private shouldProcessBatch: boolean;
+  private deferred: Deferred.Deferred<void> | undefined;
+
+  constructor(options: EventOutboxFanOutServiceOptions) {
+    this.db = options.db;
+    this.pollInterval = options.pollInterval ?? 30_000;
+    this.shouldProcessBatch = false;
+  }
+
+  async fanOutEvents(
+    params: EventOutboxFanOutService_FanOutEventsParams,
+  ): Promise<void> {
+    const { signal } = params;
+
+    const notificationPgClient = await this.db.$client.connect();
+
+    if (!notificationPgClient) {
+      throw new Error("Failed to connect to notification pg client");
+    }
+
+    const processBatchListener = () => {
+      // if there is a deferred, resolve it
+      if (this.deferred) {
+        this.deferred.resolve();
+      } else {
+        // this will be picked by the waitForNewDeliveries method
```

Comment:
`waitForNewDeliveries` -> `waitForNewEvents`?

```suggestion
        // this will be picked by the `waitForNewEvents()` method
```

Reply:
and this is for in case the while loop taking too long, that it is longer enough `processBatchListener` was triggered the 2nd+ time?